### PR TITLE
LPD-39503

### PR DESCRIPTION
**Pull request by NemethNorbert**
This is a draft to discuss the solution

# What is this trying to solve?

[LPD-39592](https://liferay.atlassian.net/browse/LPD-39503)

Because the implementation of UndoRedo relies on custom events, and we are using the onBlur event of the input fields to create new history steps, there is an edge case issue, where the user clicks on undo right after editing an input field. In this case we handle undo before we create the new history step from the blur event of the field.

# How am I fixing it?

I made sure the Undo / Redo button is disabled while we are editing an input field.

# How can you verify that it works?

Follow the reproduction steps from the ticket

# Why doesn't this include any test?

The already existing tests are sufficient, this changes the behaviour of the undo redo button by disabling it while the user is editing an input field. Because it is changing the UX, Ruth Alves approved this change

/cc @ambrinchaudhary

### CODE REVIEW

**Comment by ambrinchaudhary** [comment]
Check if this pull also solves [journalArticleAutosave.spec.ts:327:1 Undo/Redo buttons work with metadata fields ](https://testray.liferay.com/#/project/35392/routines/1021677/build/67115224/case-result/67131530 )

**Comment by ambrinchaudhary** [comment]
I have been checking and testing this issue. I do understand what the problem is and that the solution provided "solves" it. However, I am not very happy with this approach for two main reasons: 
- It is strange to me enabling/disabling Undo Redo buttons whenever the user is interacting in the Web Content edition. Not sure if this is a good idea from UX perspective (it could be, we may need to ask). 
- We are introducing more events and this is making components more unreadable, involving other components from data engine, input localized... it could be just my perception, but I am also worried because it could cause other effects. 

> we handle undo before we create the new history step from the blur event of the field.

So, I have checked that the `handleStoreState` is triggered correctly before the `handleUndo` and `handleRedo` functions, but should not be "finishing" on time (or the `handleUndo` and `handleRedo` fn start). Which makes me thing that events from outside (data engine, input localized) are triggered correctly, and we may need to find a way to "wait" until the state after the blur is correctly saved before triggering Undo or Redo actions, but all inside our UndoRedo.js component. 

I've tried with the `setTimeout` for those fn with unsuccessfully result, so we need to look for different ways. Some ideas that I haven't tested yet: 

- There is a `Liferay.once()` event that we may use. Maybe when the state after blur is saved we can trigger a `Liferay.fire()` event and wait for the `Liferay.once()` in the `handleUndo` and `handleRedo` functions. But we have to take care of multiple fires (maybe `Liferay.detach()` just before storing the step?
- Create a lock/unlock system between those functions (`handleStoreState` vs. `handleUndo` / `handleRedo`)

To be honest, I don't know if that would solve or not the issue, but I think we should investigate a little bit more. If not possible, we can go with this approach (if UX/PM are ok). I will work on this in the next days ;-)

**Comment by ambrinchaudhary** [comment]
Hi @NemethNorbert, 
I have been working on this few times but unfortunately, could not make it work using a different approach 😢. I don't like to have this stopped, so I'm ok with the solution you proposed here. Just check first with UX/PM.

**Comment by NemethNorbert** [comment]
> Hi @NemethNorbert, I have been working on this few times but unfortunately, could not make it work using a different approach 😢. I don't like to have this stopped, so I'm ok with the solution you proposed here. Just check first with UX/PM.

Hey @ambrinchaudhary ,

I approved this PR with Ruth Alves.

**Review by ambrinchaudhary — APPROVED** [review]

**Comment by ambrinchaudhary** [comment]
ci:forward
